driver-core/src/main/java/com/datastax/driver/core/DefaultResultSetFuture.java

```diff
@@ -70,6 +70,38 @@ public void onSet(
       switch (response.type) {
         case RESULT:
           Responses.Result rm = (Responses.Result) response;
+
+          if (rm.getCustomPayload() != null
+              && rm.getCustomPayload().containsKey("tablets-routing-v1")
+              && (statement instanceof BoundStatement)) {
+            BoundStatement st = (BoundStatement) statement;
+            TupleType tupleType =
+                TupleType.of(
+                    protocolVersion,
+                    session.configuration().getCodecRegistry(),
+                    DataType.bigint(),
+                    DataType.bigint(),
+                    DataType.list(
+                        TupleType.of(
+                            protocolVersion,
+                            session.configuration().getCodecRegistry(),
+                            DataType.uuid(),
+                            DataType.cint())));
+            TypeCodec<TupleValue> typeCodec =
+                session.configuration().getCodecRegistry().codecFor(tupleType);
```

Comment:
Shouldn't all this be done statically to avoid the computations? Or at least cached?

Reply:
I've moved this part to TabletMap class. Now types and codec will be initialized on first use. I'm assuming that after first negotiation the protocol version won't dynamically change.